optimizely/optimizely_config.py

```diff
@@ -136,9 +117,121 @@ def __init__(self, project_config):
         self.environment_key = project_config.environment_key
         self.attributes = project_config.attributes
         self.events = project_config.events
+        self.rollouts = project_config.rollouts
 
         self._create_lookup_maps()
 
+        '''
+            Merging typed_audiences with audiences from project_config.
+            The typed_audiences has higher presidence.
+        '''
+
+        typed_audiences = project_config.typed_audiences[:]
+        optly_typed_audiences = []
+        for typed_audience in typed_audiences:
+            optly_audience = OptimizelyAudience(
+                typed_audience.get('id'),
+                typed_audience.get('name'),
+                typed_audience.get('conditions')
+            )
+            optly_typed_audiences.append(optly_audience)
+
+        for old_audience in project_config.audiences:
+            # check if old_audience.id exists in new_audiences.id from typed_audiences
+            if len([new_audience for new_audience in project_config.typed_audiences
+                    if new_audience.get('id') == old_audience.get('id')]) == 0:
+                if old_audience.get('id') == "$opt_dummy_audience":
+                    continue
+                else:
+                    # Convert audiences lists to OptimizelyAudience array
+                    optly_audience = OptimizelyAudience(
+                        old_audience.get('id'),
+                        old_audience.get('name'),
+                        old_audience.get('conditions')
+                    )
+                    optly_typed_audiences.append(optly_audience)
+
+        self.audiences = optly_typed_audiences
+
+    def replace_ids_with_names(self, conditions, audiences_map):
+        '''
+            Gets conditions and audiences_map [id:name]
+
+            Returns:
+                a string of conditions with id's swapped with names
+                or empty string if no conditions found.
+
+        '''
+        if conditions is not None:
+            return self.stringify_conditions(conditions, audiences_map)
+        else:
+            return ''
+
+    def lookup_name_from_id(self, audience_id, audiences_map):
+        '''
+            Gets and audience ID and audiences map
+
+            Returns:
+                The name corresponding to the ID
+                or '' if not found.
+        '''
+        name = None
+        try:
+            name = audiences_map[audience_id]
+        except KeyError:
+            name = audience_id
+
+        return name
+
+    def stringify_conditions(self, conditions, audiences_map):
+        '''
+            Gets a list of conditions from an entities.Experiment
+            and an audiences_map [id:name]
+
+            Returns:
+                A string of conditions and names for the provided
+                list of conditions.
+        '''
+        ARGS = ConditionOperatorTypes.operators
```

Comment:
can we have more small comments inside this method, to be clear about each step.

Reply:
Sure, will add